### PR TITLE
Add database ratelimiting backend

### PR DESCRIPTION
**Pull request by LukasReschke**
In case no distributed memory cache is specified this adds a database backend for ratelimiting purposes. Previously the ratelimit check was just skipped.

This also fixes an issue where we passed always the current time instead of the storage period to the backend. And simplified the API to not pass `$period` where not necessary.

This doesn't aim to be a high performance solution, but just a fallback for instances without any kind of memory cache configured.

# Test plan

- Disable the memory cache in `config.php`
- Enable the "Q&A Testing App" in the app settings
- Go to `/index.php/apps/testing/userAndAnonProtected` and click a few times. Verify that the requests fails due to hitting the ratelimit.
- Verify that expired ratelimit entries are removed on requests
- Repeat the same with memory cache enabled

# TODO

- [x] Revert unnecessary composer changes
- [x] Cleanup unneeded entries on check
- [x] PHP CS fixup

### CODE REVIEW

**Comment by LukasReschke** [comment]
@skjnldsv This needs quite a careful review as it is quite a big change to backport 🙈

File: core/Migrations/Version23000Date20210906132259.php [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+namespace OC\Core\Migrations;
+
+use Closure;
+use OCP\DB\ISchemaWrapper;
+use OCP\DB\Types;
+use OCP\Migration\IOutput;
+use OCP\Migration\SimpleMigrationStep;
+
+/**
+ * Auto-generated migration step: Please modify to your needs!
```
Comment (ChristophWurst) [inline core/Migrations/Version23000Date20210906132259.php:14]: :fire:

File: lib/private/Security/RateLimiting/Backend/DatabaseBackend.php [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+namespace OC\Security\RateLimiting\Backend;
+
+use OCP\AppFramework\Utility\ITimeFactory;
+use OCP\DB\QueryBuilder\IQueryBuilder;
+use OCP\IDBConnection;
+
+/**
+ * Class DatabaseBackend uses the database for storing rate limiting data.
+ *
+ * @package OC\Security\RateLimiting\Backend
```
Comment (ChristophWurst) [inline lib/private/Security/RateLimiting/Backend/DatabaseBackend.php:35]: :fire:

File: lib/private/Security/RateLimiting/Backend/IBackend.php [hunk 3]
```diff
@@ -35,16 +35,14 @@
  */
 interface IBackend {
 	/**
-	 * Gets the amount of attempts within the last specified seconds
+	 * Gets the amount of attempts for the specified method
```
Comment (ChristophWurst) [inline lib/private/Security/RateLimiting/Backend/IBackend.php:38]:
```suggestion
	 * Gets the number of attempts for the specified method
```

if we're making things more clear :)

**Review by ChristophWurst — APPROVED** [review]

**Review by ChristophWurst — APPROVED** [review]

File: lib/private/Security/RateLimiting/Backend/DatabaseBackend.php [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+		$qb = $this->dbConnection->getQueryBuilder();
+		$qb->delete(self::TABLE_NAME)
+			->where(
+				$qb->expr()->lte('delete_after', $qb->createParameter('currentTime'))
+			)
+			->setParameter('currentTime', $currentTime, 'datetime')
+			->executeStatement();
+
+		$qb = $this->dbConnection->getQueryBuilder();
+		$qb->selectAlias($qb->createFunction('COUNT(*)'), 'count')
```
Comment (nickvergessen) [inline lib/private/Security/RateLimiting/Backend/DatabaseBackend.php:82]:
```suggestion
		$qb->select($qb->func()->count())
```

File: lib/private/Security/RateLimiting/Backend/DatabaseBackend.php [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+			->from(self::TABLE_NAME)
+			->where(
+				$qb->expr()->eq('hash', $qb->createNamedParameter($identifier, IQueryBuilder::PARAM_STR))
+			)
+			->andWhere(
+				$qb->expr()->gte('delete_after', $qb->createParameter('currentTime'))
+			)
+			->setParameter('currentTime', $currentTime, 'datetime');
```
Comment (nickvergessen) [inline lib/private/Security/RateLimiting/Backend/DatabaseBackend.php:90]:
```suggestion
			->andWhere(
				$qb->expr()->gte('delete_after', $qb->createNamedParameter($currentTime, IQueryBuilder::PARAM_DATE))
			);
```

File: lib/private/Security/RateLimiting/Backend/DatabaseBackend.php [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+		string $identifier
+	): int {
+		$currentTime = $this->timeFactory->getDateTime();
+
+		$qb = $this->dbConnection->getQueryBuilder();
+		$qb->delete(self::TABLE_NAME)
+			->where(
+				$qb->expr()->lte('delete_after', $qb->createParameter('currentTime'))
+			)
+			->setParameter('currentTime', $currentTime, 'datetime')
```
Comment (nickvergessen) [inline lib/private/Security/RateLimiting/Backend/DatabaseBackend.php:78]:
```suggestion
			->where(
				$qb->expr()->lte('delete_after', $qb->createNamedParameter($currentTime, IQueryBuilder::PARAM_DATE))
			);
```

File: core/Migrations/Version23000Date20210906132259.php [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+
+		$hasTable = $schema->hasTable(self::TABLE_NAME);
+
+		if (!$hasTable) {
+			$table = $schema->createTable(self::TABLE_NAME);
+			$table->addColumn('hash', Types::STRING, [
+				'notnull' => true,
+				'length' => 128,
+			]);
+			$table->addColumn('delete_after', 'datetime', [
```
Comment (nickvergessen) [inline core/Migrations/Version23000Date20210906132259.php:34]:
```suggestion
			$table->addColumn('delete_after', Types::DATETIME, [
```

File: core/Migrations/Version23000Date20210906132259.php [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+				'notnull' => true,
+			]);
+			$table->addIndex(['hash'], 'ratelimit_hash_idx');
+			$table->addIndex(['delete_after'], 'ratelimit_delete_after_idx');
+		}
+
+		return $schema;
```
Comment (nickvergessen) [inline core/Migrations/Version23000Date20210906132259.php:41]:
```suggestion
			return $schema;
		}

		return null;
```

**Review by nickvergessen — CHANGES_REQUESTED** [review]

**Comment by nickvergessen** [comment]
Looks good in general, just some minor changes

**Comment by LukasReschke** [comment]
Missed to set the remote. Retesting.

File: lib/private/Security/RateLimiting/Backend/DatabaseBackend.php [hunk 9]
```diff
@@ -0,0 +1,124 @@
...
+	private function getExistingAttemptCount(
+		string $identifier
+	): int {
+		$currentTime = $this->timeFactory->getDateTime();
+
+		$qb = $this->dbConnection->getQueryBuilder();
+		$qb->delete(self::TABLE_NAME)
+			->where(
+				$qb->expr()->lte('delete_after', $qb->createNamedParameter($currentTime, IQueryBuilder::PARAM_DATE))
+			);
```
Comment (nickvergessen) [inline lib/private/Security/RateLimiting/Backend/DatabaseBackend.php:77]:
```suggestion
			)
```

**Review by AndyScherzinger — APPROVED** [review]

**Pull request merged**